torch/csrc/distributed/autograd/engine/dist_engine.cpp

```diff
@@ -23,6 +24,43 @@ using torch::autograd::variable_list;
 static constexpr char* kNumBackwardPasses = "num_current_backward_passes";
 static constexpr char* kNumAutogradContexts = "num_autograd_contexts";
 
+// This hook does 2 things:
+//   1. Accumuate the gard to RPC context.
+//   2. Call post hooks of the original AccumulateGrad.
+struct DistAccumulateGradCapturePreHook
+    : GraphTask::ExecInfo::Capture::GradCapturePreHook {
+  DistAccumulateGradCapturePreHook(
+      std::shared_ptr<AccumulateGrad> accumulateGrad,
+      ContextPtr autogradContext)
+      : accumulateGrad_(std::move(accumulateGrad)),
+        autogradContext_(std::move(autogradContext)) {}
+
+  torch::Tensor operator()(const torch::Tensor& grad) override {
+    // It is possible that the grad is not defined since a separate
```

Comment:
Not sure what that comment means?
undefinted gradient means it is a Tensor full of 0s. You might want to double check the posthook logic in the engine, but I this we do call the post hooks even when gradients are undefined.

Reply:
Is there a reason we do this for the local engine? It seems like when the grad is not defined, we don't call pre-hooks and skip gradient accumulation, but still call post hooks? I'm fine with replicating the logic in the local engine, but it seems inconsistent that we only call post hooks.

Reply:
I copy-pasted the comment from the code I deleted. This is to avoid accumulating an undefined grad into the RPC context. 
@pritamdamania87 This current implementation is to skip everything if the grad is not defined, which is consistent with calling pre-hooks.


https://github.com/pytorch/pytorch/blob/c7631716da203849d15d17231db6ee0304363471/torch/csrc/distributed/autograd/engine/dist_engine.cpp#L233-L237

Reply:
We do call both pre and post hooks all the time.
Only the special variable hooks called from within the AccumulateGrad are skipped when the grad is undefined.

Reply:
I see, let's keep this consistent and still call the post hooks even if the grad is not defined.

Reply:
Ok, will call post hooks even if the grad is undefined.